service/matching/taskReader.go

```diff
@@ -308,13 +324,25 @@ func (tr *taskReader) addSingleTaskToBuffer(task *persistence.TaskInfo) bool {
 	if err != nil {
 		tr.logger.Fatal("critical bug when adding item to ackManager", tag.Error(err))
 	}
-	for {
-		select {
-		case tr.taskBuffer <- task:
-			return true
-		case <-tr.cancelCtx.Done():
-			return false
+	isolationGroup, err := tr.getIsolationGroupForTask(tr.cancelCtx, task)
+	if err != nil {
+		// it only errors when the tasklist is a sticky tasklist and
+		// the sticky pollers are not available, in this case, we just complete the task
+		// and let the decision get timed out and rescheduled to non-sticky tasklist
+		if err == _stickyPollerUnavailableError {
+			tr.completeTask(task, nil)
+		} else {
+			// it should never happen, unless there is a bug in 'getIsolationGroupForTask' method
+			tr.logger.Error("taskReader: unexpected error getting isolation group", tag.Error(err))
+			tr.completeTask(task, err)
```

Comment:
Why would we complete the task here?

Reply:
to put it back to the buffer and make sure it's not dropped